From: Tommy Webb <tommy@calyxinstitute.org>
Date: Wed, 8 Mar 2023 14:34:30 -0500
Subject: Bring Privacy Sandbox back into settings UI

Partially revert "Disable FLoC and privacy sandbox" to avert a crash.
Removing the setting can be done with policy restrictions.

Change-Id: Ifed89c20d7dda90321fe4b2bb8e5fed0bbdf9ca8
---
 .../java/res/xml/privacy_preferences.xml      |  4 +++
 .../privacy/settings/PrivacySettings.java     | 27 +++++++++++++++++++
 2 files changed, 31 insertions(+)

diff --git a/chrome/android/java/res/xml/privacy_preferences.xml b/chrome/android/java/res/xml/privacy_preferences.xml
--- a/chrome/android/java/res/xml/privacy_preferences.xml
+++ b/chrome/android/java/res/xml/privacy_preferences.xml
@@ -44,6 +44,10 @@ found in the LICENSE file.
         android:fragment="org.chromium.chrome.browser.privacy.settings.DoNotTrackSettings"
         android:key="do_not_track"
         android:title="@string/do_not_track_title"/>
+    <Preference
+        android:key="privacy_sandbox"
+        android:title="@string/prefs_privacy_sandbox"
+        android:fragment="org.chromium.chrome.browser.privacy_sandbox.PrivacySandboxSettingsFragmentV3"/>
     <PreferenceCategory
         android:key="services_category"
         android:title="@string/services_category_title">
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
--- a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
@@ -72,6 +72,7 @@ public class PrivacySettings
     private static final String PREF_SAFE_BROWSING = "safe_browsing";
     private static final String PREF_SYNC_AND_SERVICES_LINK = "sync_and_services_link";
     private static final String PREF_CLEAR_BROWSING_DATA = "clear_browsing_data";
+    private static final String PREF_PRIVACY_SANDBOX = "privacy_sandbox";
     private static final String PREF_PRIVACY_GUIDE = "privacy_guide";
     private static final String PREF_INCOGNITO_LOCK = "incognito_lock";
     private static final String PREF_THIRD_PARTY_COOKIES = "third_party_cookies";
@@ -95,6 +96,25 @@ public class PrivacySettings
                 PrivacyPreferencesManagerImpl.getInstance();
         getActivity().setTitle(R.string.prefs_privacy_security);
 
+        if (ChromeFeatureList.isEnabled(ChromeFeatureList.PRIVACY_SANDBOX_SETTINGS_4)) {
+            SettingsUtils.addPreferencesFromResource(this, R.xml.privacy_preferences_v2);
+        } else {
+            SettingsUtils.addPreferencesFromResource(this, R.xml.privacy_preferences);
+        }
+
+        Preference sandboxPreference = findPreference(PREF_PRIVACY_SANDBOX);
+        if (PrivacySandboxBridge.isPrivacySandboxRestricted()) {
+            // Hide the Privacy Sandbox if it is restricted.
+            getPreferenceScreen().removePreference(sandboxPreference);
+        } else {
+            // Overwrite the click listener to pass a correct referrer to the fragment.
+            sandboxPreference.setOnPreferenceClickListener(preference -> {
+                PrivacySandboxSettingsBaseFragment.launchPrivacySandboxSettings(getContext(),
+                        new SettingsLauncherImpl(), PrivacySandboxReferrer.PRIVACY_SETTINGS);
+                return true;
+            });
+        }
+
         Preference privacyGuidePreference = findPreference(PREF_PRIVACY_GUIDE);
         // Record the launch of PG from the S&P link-row entry point
         privacyGuidePreference.setOnPreferenceClickListener(preference -> {
@@ -273,6 +293,13 @@ public class PrivacySettings
             }
         }
 
+        Preference privacySandboxPreference = findPreference(PREF_PRIVACY_SANDBOX);
+        if (privacySandboxPreference != null
+                && !ChromeFeatureList.isEnabled(ChromeFeatureList.PRIVACY_SANDBOX_SETTINGS_4)) {
+            privacySandboxPreference.setSummary(
+                    PrivacySandboxSettingsBaseFragment.getStatusString(getContext()));
+        }
+
         mIncognitoLockSettings.updateIncognitoReauthPreferenceIfNeeded(getActivity());
 
         Preference thirdPartyCookies = findPreference(PREF_THIRD_PARTY_COOKIES);
-- 
2.40.0

